multy_core/src/bitcoin/bitcoin_transaction_segwit.cpp

```diff
@@ -0,0 +1,506 @@
+{
+    BitcoinTransactionSegWitDestinationPtr destination(
+            new BitcoinTransactionSegWitDestination(get_net_type()));
+    m_destinations.emplace_back(std::move(destination));
+
+    return register_properties(
+            make_id("#", m_destinations.size() - 1),
+            m_destinations.back()->get_properties());
+}
+
+Properties& BitcoinTransactionSegWit::get_fee()
+{
+    return m_fee->get_properties();
+}
+
+void BitcoinTransactionSegWit::set_message(const BinaryData& /*value*/)
+{
+    THROW_EXCEPTION("Not implemented yet");
+}
+
+BigInt BitcoinTransactionSegWit::estimate_total_fee(size_t /*sources_count*/, size_t /*destinations_count*/) const
+{
+    THROW_EXCEPTION("Not implemented yet");
+}
+
+BigInt BitcoinTransactionSegWit::get_total_fee() const
+{
+    return calculate_diff();
+}
+
+BigInt BitcoinTransactionSegWit::get_total_spent() const
+{
+    THROW_EXCEPTION("Not implemented yet");
+}
+
+void BitcoinTransactionSegWit::update()
+{
+    verify();
+    sign();
+
+}
+
+BinaryDataPtr BitcoinTransactionSegWit::serialize()
+{
+    update();
+
+    BitcoinDataStream data_stream;
+    data_stream << m_version;
+    data_stream << static_cast<uint8_t>(0x00); // marker
+    data_stream << static_cast<uint8_t>(0x01); // flag
+    data_stream << as_compact_size(m_sources.size());
+    for (const auto& source : m_sources)
+    {
+        data_stream << *source->serialize_with_script_pub_key();
+    }
+    data_stream << as_compact_size(m_destinations.size());
+    for (const auto& destination : m_destinations)
+    {
+        data_stream << *destination;
+    }
+
+    for (const auto& source : m_sources)
+    {
+        data_stream << *source->script_witness;
+    }
+
+    data_stream << **lock_time;
+
+    return make_clone(data_stream.get_content());
+}
+
+BitcoinNetType BitcoinTransactionSegWit::get_net_type() const
+{
+    return static_cast<BitcoinNetType>(get_blockchain_type().net_type);
+}
+
+void BitcoinTransactionSegWit::sign()
+{
+
+    for (auto& source : m_sources)
+    {
+        source->script_signature.reset();
+    }
+
+    BitcoinDataStream preview_inputs;
+    for (const auto& source : m_sources)
+    {
+        preview_inputs << *source->preview_transaction();
+    }
+    BitcoinDataStream preview_outputs;
+    for (const auto& destination : m_destinations)
+    {
+        preview_outputs << *destination;
+    }
+
+    BitcoinDataStream data;
+    for (const auto& source : m_sources)
+    {
+        data << **source->sequence;
+    }
+
+
+    hash<256> hash_prevouts = do_hash<SHA2_DOUBLE, 256>(preview_inputs.get_content());
+    hash<256> hash_sequence = do_hash<SHA2_DOUBLE, 256>(data.get_content());
+    hash<256> hash_outpusts = do_hash<SHA2_DOUBLE, 256>(preview_outputs.get_content());
+
+    for (auto& source : m_sources)
+    {
+        source->sign(hash_prevouts, hash_sequence, hash_outpusts, *lock_time);
+    }
+
+
+}
+
+void BitcoinTransactionSegWit::verify()
```

Comment:
This looks like 99% copy of `BitcoinTransaction::verify()`. please move common code to base class.